### DISCUSSION TITLE
How does Riallto compile C/C++ into xclbin?

### DISCUSSION THREAD

**Discussion by marty1885** [AMDResearch/Riallto#24] [Q&A] [answered]
HI,

I'm looking into using the IPU for my own project. I can't use Riallto directly due to need of calling from C++. I've looked into how the RyzenAI stack works and thinks Riallto most likely goes through MLIR-AIE. However, I don't get how Riallto is able to accept kernels as C++ source code. AFAIK MLIR-AIE only accepts MLIR as input.

How does Riallto compile C/C++ into MLIR? Is there an open tool that can do it?

**Comment by STFleming** [comment, marked as answer]
Hi @marty1885,

In Riallto we use small subset of the Vitis tools for compiling the C++ kernel code. This kernel code gets compiled into an object file that we can then call on an AIE-Tile from within MLIR-AIE. An example of this can be seen in the repo [here](https://github.com/AMDResearch/Riallto/blob/c20c6e1a630c77527f43b2e8e17434634cce71de/npu/lib/applications/binaries/edge_detect_720p.mlir#L68). 

All the best,
Shane

**Reply by marty1885** [reply]
Cool thanks! A follow up question. How does Riallto know how much arguments a kernel as/how to assign them? Also is there guides on how to write the MLIR side code by hand (beyond the tutorials and speech slides)?

**Reply by STFleming** [reply]
No problem at all. The Cpp kernel source is parsed to figure out the types/names [here](https://github.com/AMDResearch/Riallto/blob/c20c6e1a630c77527f43b2e8e17434634cce71de/npu/build/kernel.py#L67). There are some [tutorials](https://github.com/Xilinx/mlir-aie/tree/main/mlir_tutorials) on the MLIR-AIE repo that might be helpful for writing the MLIR-AIE code directly.
